### ISSUE TITLE
 TensorFlow Lite GPU on Android slower than CPU

### ISSUE THREAD

**Issue by EarlHsy** [tensorflow/tensorflow#31826]
**System information**
Android 9.0, XiaoMi9, snapdragon855
- TensorFlow installed from (source or binary):
https://github.com/tensorflow/tensorflow/tree/master/tensorflow/lite/java/demo
- TensorFlow version (or github SHA if from source):
dependencies {
    ...
    implementation 'org.tensorflow:tensorflow-lite:0.0.0-nightly'
    implementation 'org.tensorflow:tensorflow-lite-gpu:0.0.0-nightly'
}

I run a app from https://github.com/tensorflow/tensorflow/tree/master/tensorflow/lite/java/demo, and it worked well, the run time of  default model(mobilenet) on GPU is half of CPU, which is normal I think. And I replace the model with a special model which structure like YOLOv2-Tiny, trained by keras and convert to tflite(Implemented by traditional conv2D and pooling), I found the speed on GPU is slower than CPU float(CPU is less than 500ms but GPU is more than 600ms),  is it normal?
is there anyone help me?

**Comment by wangyiwei0306** [comment]
Have you sloved your problem? I also encountered the same problem. The model I use is YOLOv3 and also trained by keras.

**Comment by EarlHsy** [comment]
> Have you sloved your problem? I also encountered the same problem. The model I use is YOLOv3 and also trained by keras.

No i'm not solved this problem yet, and i don't have so many Android phones to test this. I tested in a xiaomi 5 with snapdragan 820 SoC and the problem still remained. I think right now the best way is waiting and hope tensorflow lite will fix this problem in the future.

**Comment by impjdi** [comment]
OMG this issue completely dropped off my radar.  Didn't even know this was assigned to me.  My apologies.

I'm not familiar with YOLO, so would be great if you can share your YOLOv2-Tiny TFLite model somehow.  Either the entire network is not compatible with TFLite GPU (unsupported ops) and thus only parts of it runs on the GPU and the rest on the CPU.  This can cause in a significant performance loss.

Also, if YOLO employs FC layers, performance can suffer, because FC on OpenGL is not that fast, and sometimes slower than CPU.

**Comment by EarlHsy** [comment]
> OMG this issue completely dropped off my radar. Didn't even know this was assigned to me. My apologies.
> 
> I'm not familiar with YOLO, so would be great if you can share your YOLOv2-Tiny TFLite model somehow. Either the entire network is not compatible with TFLite GPU (unsupported ops) and thus only parts of it runs on the GPU and the rest on the CPU. This can cause in a significant performance loss.
> 
> Also, if YOLO employs FC layers, performance can suffer, because FC on OpenGL is not that fast, and sometimes slower than CPU.

I‘m sorry that i don't have this model in my computer right now, but i do remember YOLOv2-Tiny consists of maxpooling and convolution layers, no FCs, so I dont know why. I'll do more tests.

**Comment by wangyiwei0306** [comment]
@EarlHsy Hi， I encountered another question. I found that the detection results are also different. I checked the outputmap of function "runForMultipleInputsOutputs" and found that the values in the outputmap are also totally different. Do you have the same problem,too?

**Comment by sushreebarsa** [comment]
@EarlHsy Could you please check whether your model is falling back to CPU or not. Please confirm if the model is utilizing the GPU fully.
Thank you!
